### ISSUE TITLE
setup upgrade command will purge whole full page cache (varnish)

### ISSUE THREAD

**Issue by onlinebizsoft** [magento/magento2#37336]
### Preconditions and environment

- Magento version 2.4.5 (but should happen with any version)

### Steps to reproduce

1. Run command setup:upgrade
2. Validate cached pages in varnish

### Expected result

Cached pages should stay in varnish

### Actual result

Whole varnish were purged

### Additional information

/vendor/magento/framework/App/Cache/Manager.php : get all cache type to clean

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by onlinebizsoft** [comment]
Do you have any idea about this behavior? @hostep @ihor-sviziev

**Comment by ihor-sviziev** [comment]
I think it's designed like that because expecting this command to be executed during deployment. And when you deploy a new build, you probably might deploy some new module, new magento version, new css or js. And after the deployment you expect to see the change you deployed.
I do agree that not all changes require varnish cache invalidation, but developing mechanism that will decide if we need to flush cache or no - pretty hard.
From my perspective, we should keep such behavior as expected

**Comment by onlinebizsoft** [comment]
@ihor-sviziev but purging varnish can be done at very last step because after upgrading DB we will have di compile and deploy static steps which can take few hours for a system with so many languages (even ran multiple processes for deploying static).
And the strange thing is that Magento clear varnish at that time but after finishing whole the deployment processes, we still have to clear varnish again. So what is the point doing it twice?

**Comment by ihor-sviziev** [comment]
@onlinebizsoft for production environment it should be the last step.
Here are the docs how it should be
https://devdocs.magento.com/guides/v2.3/config-guide/deployment/pipeline/technical-details.html#config-deploy-workflow

**Comment by hostep** [comment]
@onlinebizsoft: Ihor is correct, `setup:upgrade` (preferably with the ` --keep-generated` flag) should be one of the last steps of the deployment process, `setup:di:compile` and `setup:static-content:deploy` should happen before that.
And it's expected to flush the entire cache.

> ... and deploy static steps which can take few hours for a system with so many languages (even ran multiple processes for deploying static).

Multiple **hours**? That must surely mean there is something wrong, because that's insane. The longest deploy we have with a project takes 11 minutes and deploys 14 languages using 3 parallel jobs.
You might be running against [this bug](https://github.com/magento/magento2/issues/36185), which was solved by https://github.com/magento/magento2/search?q=ACP2E-1168&type=commits

**Comment by onlinebizsoft** [comment]
@hostep our project has more than 60 languages so it took quite some time. I will look at deployment steps again, look like we shouldnt run upgrade before static.

**Comment by engcom-November** [comment]
Hi @onlinebizsoft ,
We are closing this issue as there has been no latest update on the same. Kindly reopen / create new issue if you are still facing issues.
Thank you.

**Comment by onlinebizsoft** [comment]
@engcom-November what do you mean by "no latest update on the same" ? Please re-open, this issue is still valid

**Comment by engcom-Dash** [comment]
Hi @onlinebizsoft ,

Verified the issue in 2.4-develop instance and the issue is not reproducible.

Steps to reproduce:
1.Install magento instance
2.Run Setup Upgrade command
3.Validate cached pages in varnish as per [Devedocs](https://experienceleague.adobe.com/docs/commerce-operations/configuration-guide/cache/config-varnish-final.html?lang=en)

Kindly refer the below screenshots:
<img width="1505" alt="Screenshot 2023-06-14 at 5 10 40 PM" src="https://github.com/magento/magento2/assets/60198592/5c47d75c-cbb0-47da-9824-e0f5cbd1bb78">
<img width="1505" alt="Screenshot 2023-06-14 at 5 11 05 PM" src="https://github.com/magento/magento2/assets/60198592/0622c23f-1544-434a-ae95-a71dc26f7d13">

After upgrading cached pages stay in varnish only.

Let us know if we are missing anything!

**Comment by hostep** [comment]
@engcom-Dash: there's no good reason to put time in this issue.

@onlinebizsoft : I'm afraid Magento is designed like this to flush all caches each time `bin/magento setup:upgrade` runs, I don't think it's a good idea to attempt to change that, as it would probably break thousands of different workflows people are now used to use.

In my opinion this ticket should be closed again.

**Comment by onlinebizsoft** [comment]
@hostep yes, I do agree that upgrade should be done at the last step however sometime we may deploy some new backend modules,....which should run upgrade DB but shouldn't touch Varnish.
Automatic detection can be hard but maybe we should provide an option to flush Varnish or not when we run upgrade DB so the deployment pipeline can decide which strategy they want

**Comment by ihor-sviziev** [comment]
@onlinebizsoft I think it's not so important to provide such a feature out of the box since it would be needed for small percent of shops.
If needed, you always could implement it in your custom extension or apply your custom patch

**Comment by engcom-Dash** [comment]
Hi @onlinebizsoft ,
 
As per @hostep [Comments](https://github.com/magento/magento2/issues/37336#issuecomment-1591406118) ,We are closing the issue for now. Please feel free to reopen it or ask us to reopen.

Thanks!
